### PR TITLE
feat: Add referenced table metadata to namingstrategy to resolve foreignKeyName

### PR DESCRIPTION
**Pull request by tkvw**
@see https://github.com/typeorm/typeorm/issues/3847 and https://github.com/typeorm/typeorm/issues/1355

### CODE REVIEW

File: src/naming-strategy/NamingStrategyInterface.ts [hunk 1]
```diff
@@ -60,7 +60,7 @@ export interface NamingStrategyInterface {
     /**
      * Gets the name of the foreign key.
      */
-    foreignKeyName(tableOrName: Table|string, columnNames: string[]): string;
+    foreignKeyName(tableOrName: Table|string, columnNames: string[], referencedTablePath: string, referencedColumnNames: string[]): string;
```
Comment (pleerock) [inline src/naming-strategy/NamingStrategyInterface.ts]: we need to mark them with `?` to prevent breaking release

**Comment by pleerock** [comment]
Also would be great to have a test for this use case :) Also please add a PR title according to our standards. Thanks :)

**Comment by tkvw** [comment]
Fixed conversation, rebased commit and added test

**Comment by kedar-moghe** [comment]
Will this give the ability to give our own custom constraint names, or will it be generated by the scheme in the code?

**Comment by tkvw** [comment]
> Will this give the ability in the future to give our own custom constraint names, or will it just be generated by the scheme in the code?

No, that is already possible using the namingstrategy. This PR just add some additional arguments to generate a foreignkey name based on referenced columns. This is not possible at te moment.

**Comment by pleerock** [comment]
merged, sorry for late response here

**Comment by tdnghia98** [comment]
> namingstrategy

Is there any support for this? It could be really useful when debugging, having generic constraints name is confusing

**Comment by FernandoRicardez** [comment]
So it is posible to change the FK_0669fe20e252eb692bf4d344975, To something custom or at least more readable if it is some documentation would be neat!

**Comment by kaykhancheckpoint** [comment]
For those struggling with this, it is possible to explicitly set the name of FK constraints by building a naming strategy.

**CustomNamingStrategy**

The following naming convention is used below: `<table1>_<table2>_<table1_column>` then i hash this because it could get long and mysql will throw an identifier too long error.

This method makes the foreign keys deterministic so that i can use the same migrations for both test and production db.

```
import { DefaultNamingStrategy, Table, NamingStrategyInterface } from "typeorm";
import crypto from "crypto";

export class CustomNamingStrategy extends DefaultNamingStrategy implements NamingStrategyInterface {
    foreignKeyName(tableOrName: Table | string, columnNames: string[], referencedTablePath?: string, referencedColumnNames?: string[]): string {
        
        tableOrName =
            typeof tableOrName === "string" ? tableOrName : tableOrName.name;

        const name = columnNames.reduce(
            (name, column) => `${name}_${column}`,
            `${tableOrName}_${referencedTablePath}`,
        );

        return`fk_${crypto.createHash('md5').update(name).digest("hex")}`
    }
}
```


**ormconfig.ts**

Then import the custom naming strategy in your config.

```
module.exports = {
    "type": "mysql",
    "host": process.env.DB_HOST,
    ...
    namingStrategy: new CustomNamingStrategy()
}
```

references:

https://github.com/typeorm/typeorm/tree/master/sample/sample12-custom-naming-strategy
https://github.com/typeorm/typeorm/issues/5286
https://github.com/typeorm/typeorm/pull/4274/files

**Comment by emtiajium** [comment]
> For those struggling with this, it is possible to explicitly set the name of FK constraints by building a naming strategy.
> 
> **CustomNamingStrategy**
> 
> The following naming convention is used below: `<table1>_<table2>_<table1_column>` then i hash this because it could get long and mysql will throw an identifier too long error.
> 
> This method makes the foreign keys deterministic so that i can use the same migrations for both test and production db.
> 
> ```
> import { DefaultNamingStrategy, Table, NamingStrategyInterface } from "typeorm";
> import crypto from "crypto";
> 
> export class CustomNamingStrategy extends DefaultNamingStrategy implements NamingStrategyInterface {
>     foreignKeyName(tableOrName: Table | string, columnNames: string[], referencedTablePath?: string, referencedColumnNames?: string[]): string {
>         
>         tableOrName =
>             typeof tableOrName === "string" ? tableOrName : tableOrName.name;
> 
>         const name = columnNames.reduce(
>             (name, column) => `${name}_${column}`,
>             `${tableOrName}_${referencedTablePath}`,
>         );
> 
>         return`fk_${crypto.createHash('md5').update(name).digest("hex")}`
>     }
> }
> ```
> 
> **ormconfig.ts**
> 
> Then import the custom naming strategy in your config.
> 
> ```
> module.exports = {
>     "type": "mysql",
>     "host": process.env.DB_HOST,
>     ...
>     namingStrategy: new CustomNamingStrategy()
> }
> ```
> 
> references:
> 
> https://github.com/typeorm/typeorm/tree/master/sample/sample12-custom-naming-strategy #5286 https://github.com/typeorm/typeorm/pull/4274/files

Which TypeORM version are you using? I use v0.2.32, and it simply ignores my custom naming strategy. 

**My codes**
https://github.com/emtiajium/ielts-gibberish/blob/b5717f4a52b9f24ead1e11fc9220f1c3e127b4c0/src/common/persistence/DatabaseModule.ts#L46

https://github.com/emtiajium/ielts-gibberish/blob/0e8703b52dc8dd8fd7a4c17bd0eb7f63b49c838e/src/common/persistence/DatabaseNamingStrategy.ts#L16

**Pull request merged**